### PR TITLE
simplify and improve type intersection algorithm a bit

### PR DESCRIPTION
**Pull request by JeffBezanson**
This removes some code and makes a class of results more conservative, fixing some potential cases of unsoundness.

I stumbled into this while working on #41738. It doesn't fix that issue yet but should make it easier, and was a nice discovery in the process. See the bottom of the diff for a changed test case where I'm pretty sure we were getting the wrong answer before. Let's see if this passes CI.

### CODE REVIEW

**Comment by JeffBezanson** [comment]
@nanosoldier `runtests(ALL, vs = ":master")`

**Review by vtjnash — APPROVED** [review]

**Comment by nanosoldier** [comment]
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/41795#issuecomment-893964479) has completed - possible new issues were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/pkgeval/by_hash/0f389b0_vs_a86fe0f/report.md).

File: test/subtype.jl [hunk 1]
```diff
@@ -1813,7 +1821,7 @@ let
     # issue #38279
     t = typeintersect(Tuple{<:Array{T, N}, Val{T}} where {T<:Real, N},
                       Tuple{<:Array{T, N}, Val{<:AbstractString}}  where {T<:Real, N})
-    @test t == Tuple{<:Array{Union{}, N}, Val{Union{}}} where N
+    @test t == Tuple{<:Array{T, N}, Val{Union{}}} where {T<:Real, N}
```
Comment (martinholters) [inline test/subtype.jl:1824]: Could you explain why the answer was wrong before? That means that `Tuple{<:Array{Union{}, N}, Val{Union{}}} where N`is too tight, right? So there is a type that is a subtype of both intersected types but not of the old result? I don't see it. Can you give an example?

Also, given that `(Tuple{<:Array{T, N}, Val{T}} where {T<:Real, N}) == (Tuple{Array{T, N}, Val{T}} where {T<:Real, N})` and `(Tuple{<:Array{T, N}, Val{<:AbstractString}}  where {T<:Real, N}) == (Tuple{Array{T, N}, Val{<:AbstractString}}  where {T<:Real, N})`, this change equally affects `typeintersect(Tuple{Array{T, N}, Val{T}} where {T<:Real, N}, Tuple{Array{T, N}, Val{<:AbstractString}}  where {T<:Real, N})`, right?

Comment (JeffBezanson) [reply test/subtype.jl:1824]: Yeah I guess this wasn't wrong before after all. And in fact if you remove the redundant typevars, this still gives
```
Tuple{Array{Union{}, N}, Val{Union{}}} where N
```

**Comment by JeffBezanson** [comment]
Looks like there are some new stack overflows in package tests.

**Comment by JeffBezanson** [comment]
At this point I have a change that *just barely* fixes #41738, by managing to return a non-concrete type, but one that is still too narrow. There are a couple minor test changes that I think are reasonable. I have an idea for how to get a more correct result, but we may be able to use this in a pinch.

**Comment by JeffBezanson** [comment]
@nanosoldier `runtests(ALL, vs = ":master")`

**Comment by nanosoldier** [comment]
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/41795#issuecomment-895405803) has completed - possible new issues were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/pkgeval/by_hash/f3ec252_vs_afc7e72/report.md).

**Comment by JeffBezanson** [comment]
Stack overflow in ValkyrieRobot.

**Comment by JeffBezanson** [comment]
Modified to use even less stack space.

**Comment by nanosoldier** [comment]
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/41795#issuecomment-896385382) has completed - possible new issues were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/pkgeval/by_hash/fc98aec_vs_cc3fc0b/report.md).

**Comment by JeffBezanson** [comment]
@nanosoldier `runtests(ALL, vs = ":master")`

**Comment by nanosoldier** [comment]
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/41795#issuecomment-897909341) has completed - possible new issues were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/pkgeval/by_hash/fcdb63d_vs_01d439e/report.md).

**Comment by JeffBezanson** [comment]
SumOfSquares has an unreachable-reached; other than that we're getting there.

**Comment by JeffBezanson** [comment]
@nanosoldier `runtests(ALL, vs = ":master")`

**Comment by nanosoldier** [comment]
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/41795#issuecomment-899089047) has completed - possible new issues were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/pkgeval/by_hash/e86099b_vs_b40ae6b/report.md).

**Comment by JeffBezanson** [comment]
@nanosoldier `runtests(["AdvancedVI", "BifurcationInference", "BifurcationKit", "BlackBoxOptim", "Causal", "ConstructionBase", "Dance", "DelayDiffEq", "DiffEqParamEstim", "DynamicBoundsBase", "Expect", "FMIFlux", "ForwardDiff", "LazyAlgebra", "MCPhyloTree", "Minio", "ODE", "PowerGraphics", "RemoveLFS", "SalesForceBulkApi", "SampledSignals", "SpatialJackknife", "StochasticRounding", "SymbolicRegression", "YAActL"], vs = ":master")`

**Comment by nanosoldier** [comment]
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/41795#issuecomment-899629424) has completed - possible new issues were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/pkgeval/by_hash/e86099b_vs_b40ae6b/report.md).

**Comment by JeffBezanson** [comment]
@nanosoldier `runtests(["AdvancedVI", "BifurcationInference", "BifurcationKit", "BlackBoxOptim", "Causal", "ConstructionBase", "Dance", "DelayDiffEq", "DiffEqParamEstim", "DynamicBoundsBase", "Expect", "FMIFlux", "ForwardDiff", "LazyAlgebra", "MCPhyloTree", "Minio", "ODE", "PowerGraphics", "RemoveLFS", "SalesForceBulkApi", "SampledSignals", "SpatialJackknife", "StochasticRounding", "SymbolicRegression", "YAActL"], vs = ":master")`

**Comment by nanosoldier** [comment]
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/41795#issuecomment-899843558) has completed - possible new issues were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/pkgeval/by_hash/bc00546_vs_d612c47/report.md).

**Comment by JeffBezanson** [comment]
The remaining problem seems to be ConstructionBase, where intersection is hanging or possibly just taking a very long time.

**Comment by JeffBezanson** [comment]
The slowdown happens with just b06f8133d9049dcb893db10f376f63efc6841781, to narrow it down a bit.

File: base/compiler/utilities.jl [hunk 2]
```diff
@@ -156,7 +156,11 @@ function subst_trivial_bounds(@nospecialize(atypes))
     end
     v = atypes.var
     if isconcretetype(v.ub) || v.lb === v.ub
-        return subst_trivial_bounds(atypes{v.ub})
+        try
+            return subst_trivial_bounds(atypes{v.ub})
```
Comment (vtjnash) [inline base/compiler/utilities.jl:160]: `subst_trivial_bounds` should not be in the try/catch

**Comment by JeffBezanson** [comment]
@nanosoldier `runtests(["Causal", "ConstructionBase", "DiffEqParamEstim", "DynamicBoundsBase", "FMIFlux", "SpatialJackknife"], vs = ":master")`

**Comment by nanosoldier** [comment]
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/41795#issuecomment-904133387) has completed - no new issues were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/pkgeval/by_hash/0a1bf84_vs_292f1a9/report.md).

**Comment by JeffBezanson** [comment]
@nanosoldier `runtests(ALL, vs = ":master")`

**Comment by nanosoldier** [comment]
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/41795#issuecomment-904288023) has completed - possible new issues were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/pkgeval/by_hash/0a1bf84_vs_292f1a9/report.md).

**Comment by JeffBezanson** [comment]
Looking good. Just to double check:

@nanosoldier `runtests(["CMAEvolutionStrategy", "ForwardDiff", "IntervalTrees", "JetPackDSP", "LoopThrottle", "MatrixNetworks", "SparseRegression", "SpatialJackknife", "Tectonic"], vs = ":master")`

**Comment by nanosoldier** [comment]
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/41795#issuecomment-904735658) has completed - no new issues were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/pkgeval/by_hash/0a1bf84_vs_3d058b2/report.md).

**Comment by JeffBezanson** [comment]
Note: requires #41867 and #41976

**Pull request merged**